File: src/2023/day1/day1.py
from utils.load_input import load_lines
import re

is_num = re.compile(r"\d")

words = ["zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

def parse(s: str) -> str:
    print(s)
    nums = []

    def replace(s: str) -> str:
        if is_num.match(s[0]):
            nums.append(int(s[0]))
            return s[1:]
        for i, word in enumerate(words):
            if s.startswith(word):
                nums.append(i)
                return s[len(word):]
        return s[1:]

    while len(s) > 0:
        s = replace(s)

    n = int(f"{nums[0]}{nums[-1]}")
    print(n)
    return n

sum = 0
for line in load_lines(__file__):
    sum += parse(line)

print(sum)

File: Sources/2023/day1/day1.swift
import Foundation

struct Day1 {
  static let strs = ["zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
  static let nums = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]

  static func main() throws {
    let input = try String(contentsOfFile: "Resources/2023/day1/input.txt", encoding: .utf8)
    let lines = input.split(separator: "\n")
    var sum = 0
    for line in lines {
      sum += parseStart(line) * 10 + parseEnd(line)
    }
    print(sum)
  }

  static func parseStart(_ s: String.SubSequence) -> Int {
    var s = s
    while true {
      for n in nums {
        if s.hasPrefix(n) {
          return nums.firstIndex(of: n)!
        }
      }
      for n in strs {
        if s.hasPrefix(n) {
          return strs.firstIndex(of: n)!
        }
      }
      s = s.dropFirst()
      if s.isEmpty {
        return 0
      }
    }
  }

  static func parseEnd(_ s: String.SubSequence) -> Int {
    var s = s
    while true {
      for n in nums {
        if s.hasSuffix(n) {
          return nums.firstIndex(of: n)!
        }
      }
      for n in strs {
        if s.hasSuffix(n) {
          return strs.firstIndex(of: n)!
        }
      }
      s = s.dropLast()
      if s.isEmpty {
        return 0
      }
    }
  }
}
